motionplan/plan.go

```diff
@@ -0,0 +1,243 @@
+
+// Trajectory is a slice of maps describing a series of Inputs for a robot to travel to in the course of following a Plan.
+// Each item in this slice maps a Frame's name (found by calling frame.Name()) to the Inputs that Frame should be modified by.
+type Trajectory []map[string][]referenceframe.Input
+
+// GetFrameInputs is a helper function which will extract the waypoints of a single frame from the map output of a trajectory.
+func (traj Trajectory) GetFrameInputs(frameName string) ([][]referenceframe.Input, error) {
+	solution := make([][]referenceframe.Input, 0, len(traj))
+	for _, step := range traj {
+		frameStep, ok := step[frameName]
+		if !ok {
+			return nil, fmt.Errorf("frame named %s not found in trajectory", frameName)
+		}
+		solution = append(solution, frameStep)
+	}
+	return solution, nil
+}
+
+// String returns a human-readable version of the trajectory, suitable for debugging.
+func (traj Trajectory) String() string {
+	var str string
+	for _, step := range traj {
+		str += "\n"
+		for frame, input := range step {
+			if len(input) > 0 {
+				str += fmt.Sprintf("%s: %v\t", frame, input)
+			}
+		}
+	}
+	return str
+}
+
+// EvaluateCost calculates a cost to a trajectory as measured by the given distFunc Metric.
+func (traj Trajectory) EvaluateCost(distFunc ik.SegmentMetric) float64 {
+	var totalCost float64
+	last := map[string][]referenceframe.Input{}
+	for _, step := range traj {
+		for frame, inputs := range step {
+			if len(inputs) > 0 {
+				if lastInputs, ok := last[frame]; ok {
+					cost := distFunc(&ik.Segment{StartConfiguration: lastInputs, EndConfiguration: inputs})
+					totalCost += cost
+				}
+				last[frame] = inputs
+			}
+		}
+	}
+	return totalCost
+}
+
+// Path is a slice of PathSteps describing a series of Poses for a robot to travel to in the course of following a Plan.
+type Path []PathStep
+
+func newPath(solution []node, sf *solverFrame) (Path, error) {
+	path := make(Path, 0, len(solution))
+	for _, step := range solution {
+		inputMap := sf.sliceToMap(step.Q())
+		poseMap := make(map[string]*referenceframe.PoseInFrame)
+		for frame := range inputMap {
+			tf, err := sf.fss.Transform(inputMap, referenceframe.NewPoseInFrame(frame, spatialmath.NewZeroPose()), referenceframe.World)
+			if err != nil {
+				return nil, err
+			}
+			pose, ok := tf.(*referenceframe.PoseInFrame)
+			if !ok {
+				return nil, errors.New("pose not transformable")
+			}
+			poseMap[frame] = pose
+		}
+		path = append(path, poseMap)
+	}
+	return path, nil
+}
+
+func newPathFromRelativePath(path Path) (Path, error) {
+	if len(path) < 2 {
+		return nil, errors.New("need to have at least 2 elements in path")
+	}
+	newPath := make([]PathStep, 0, len(path))
+	newPath = append(newPath, path[0])
+	for i, step := range path[1:] {
+		newStep := make(PathStep, len(step))
+		for frame, pose := range step {
+			newStep[frame] = referenceframe.NewPoseInFrame(referenceframe.World, spatialmath.Compose(newPath[i][frame].Pose(), pose.Pose()))
```

Comment:
took me a while to figure out this was looking up the previously computed pose, may I request the following minor change?:
```suggestion
		for frame, pose := range step {
		        lastPose := newPath[i][frame].Pose()
			newStep[frame] = referenceframe.NewPoseInFrame(referenceframe.World, spatialmath.Compose(lastPose, pose.Pose()))
```

Reply:
yep will change